### ISSUE TITLE
Requirement python = "^3.8" conflicts with python ">=3.8,<3.11" even though 3.11 has not been released

### ISSUE THREAD

**Issue by couling** [python-poetry/poetry#5798]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: Ubuntu 22.04 (Jammy)
- **Poetry version**: 1.1.13
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: _todo, need to strip it down for commercial reasons_

### High level

Poetry incorrectly assumes that a package specifying `python = ">=3.8,<3.11"` is incompatible with our package specifying `python = "^3.8"`.  

Poetry is interpreting `^3.8` as `>=3.8,<4.0`.  As far as I know [3.11 has not been released](https://docs.python.org/3.11/whatsnew/3.11.html) yet (2022-06-08).  So there is no released version of python that is not both `<4.0` and also `<3.11`.  [It's certainly not installed on the system](https://packages.ubuntu.com/jammy/python3).  But poetry is blindly assuming it is and so incorrectly identifying a dependency conflict.

### Details

`poetry update` is incorrectly holding back package scipy to version 1.8.1 and instead using 1.6.1 for invalid reasons.  This is problematic for us because scipy 1.6.1 has no published wheel compatible with Ubuntu 22.04 and our CI builds are now taking over 20 minutes trying to build it.

When we tried pinning scipy using:
```
scipy = "1.8.1"
```
We got this error:

```
Updating dependencies
Resolving dependencies... (1.8s)

  SolverProblemError

  The current project's Python requirement (>=3.8,<4.0) is not compatible with some of the required packages Python requirement:
    - scipy requires Python >=3.8,<3.11, so it will not be satisfied for Python >=3.11,<4.0
  
  Because scipy (1.8.1) requires Python >=3.8,<3.11
   and no versions of scipy match >1.8.1,<2.0.0, scipy is forbidden.
  So, because our-own-package depends on scipy (^1.8.1), version solving failed.

  at ~/.poetry/lib/poetry/puzzle/solver.py:241 in _solve
      237│             packages = result.packages
      238│         except OverrideNeeded as e:
      239│             return self.solve_in_compatibility_mode(e.overrides, use_latest=use_latest)
      240│         except SolveFailure as e:
    → 241│             raise SolverProblemError(e)
      242│ 
      243│         results = dict(
      244│             depth_first_search(
      245│                 PackageNode(self._package, packages), aggregate_package_nodes

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties
    
    For scipy, a possible solution would be to set the `python` property to ">=3.8,<3.11"

    https://python-poetry.org/docs/dependency-specification/#python-restricted-dependencies,
    https://python-poetry.org/docs/dependency-specification/#using-environment-markers
```

**Comment by Secrus** [comment]
This is not incorrect. That is exactly how the `^` mark is meant to be interpreted.

**Comment by couling** [comment]
@Secrus There is no actual set of **released** versions that conflicts.  So I think this is ultimately a bug (or critically problematic limitation) in the solver.

Note I do understand that working with actual sets of available versions causes the computational complexity to explode.

Perhaps what's required to make this "work" is to run a first pass on all max / min constraints and tighten them to the max and min of actual released versions...

So `<4.0` becomes `<3.11` because no `3.11` actually exists.

**Comment by couling** [comment]
Hi @neersighted I'm drawing blanks on google, so documentation links are welcome. 

My concerns here are quite specific, Poetry's (unusual) behaviour is causing problems:

 1. **I shouldn't need to declare that my wheel is incompatible with python 3.11.  It's not.**
    - Declared dependencies are written into published package wheels (are they not?).  
    - It's only one dependency that cannot install on python 3.11.  Figuring that out is the job of downstream package managers when they try to install my package.
    - I shouldn't need to re-build and publish my wheel when a new version of scipy becomes compatible with python 3.11. 
  Thats what the `^` and `~`  constraints are for in the first place.
 
 2. **How am I supposed to (CI) test with the latest versions of dependences?**
    - Poetry is refusing to install these where pip (and others) wouldn't.
    - I can't dictate to downstream users which package manager to use.
    - a virtual environment containing the current lates packages of all three is not invalid (python 3.10.5, my package, scipy 1.8.1).  No dependency constraint has been broken.

 3. Following the advice given in the error output will trick developers into adding phantom constraints to their wheels.  That's going to do nothing but fuel dependency hell for others.
 
------------


Without finding documentation to explain it, I'm having trouble understanding what Poetry is trying to achieve here:

 - Is it a case that `poetry update` is trying to ensure the lock file will work with all declared (and future) versions of python?
 - Does it do this with just python or does it do it with all dependency constraints?

Based on @neersighted 's comments I suspect that Poetry is interpreting dependency constraints as positive declarations of compatibility, and no other package manager that I'm aware of does this - for any language.  If positive declarations of compatibility are desirable then maybe they should be a different section in `pyproject.toml` so they can be maintained separately from the dependencies declared in the final (built) package.

**Comment by couling** [comment]
@neersighted I just re-read your comment and missed something important

>  I'm not sure what the ask is here -- Poetry's solver knows nothing about released vs hypothetical/future version -- it only knows about version constraints and **available distributions** ... and cannot reason about anything past that

That's not true.  Here's why (error message above):

```
The current project's Python requirement (>=3.8,<4.0) is not compatible with some of the required packages Python requirement:
    - scipy requires Python >=3.8,<3.11, so it will not be satisfied for Python >=3.11,<4.0
```

The above logic is not taking into account "available distributions" of python.  The highest available distribution of python is `3.10.5` and the only installed version on the system is `3.10.4`.  So the above error about `<3.11` and `<4.0` constraints has nothing to do with **available distributions** at all...

...The logic that's applied above only makes sense when you assume some hypothetical/future version of python that's `>=3.11,<4.0`.  But no such version is **available**.  So Poetry's solver is, in part, derived from logic based on hypothetical future versions.

Ironically what I'm asking for here is what you describe it's already doing.
